File: ts/test/c2020_12/c2020_12_15.ts
import {expect} from 'chai'

function sortedSquares(nums: number[]): number[] {
    if (nums.length == 1 || nums[0] >= 0) {
        return nums.map(n => n * n);
    } else {
        let l = 0;
        let r = nums.length - 1;
        const result: number[] = Array(nums.length).fill(0);

        while (l <= r) {
            if (-nums[l] > nums[r]) {
                result[r - l] = nums[l] * nums[l];
                l++;
            } else {
                result[r - l] = nums[r] * nums[r];
                r--;
            }
        }

        return result;
    }
}

// noinspection JSUnusedLocalSymbols
function sortedSquaresMapSort(nums: number[]): number[] {
    nums = nums.map(n => n * n);
    nums.sort((a, b) => a - b);

    return nums;
}

describe('c2020_12_15 Squares of a Sorted Array', () => {
    it('Example 1: (nums = [-4,-1,0,3,10]) -> [0,1,9,16,100]', () => {
        expect(sortedSquares([-4, -1, 0, 3, 10])).to.eql([0, 1, 9, 16, 100]);
        //Explanation: After squaring, the array becomes [16,1,0,9,100].
        //After sorting, it becomes [0,1,9,16,100].
    })
    it('Example 2: (nums = [-7,-3,2,3,11]) -> [4,9,9,49,121]', () => {
        expect(sortedSquares([-7, -3, 2, 3, 11])).to.eql([4, 9, 9, 49, 121]);
    })

    it('(nums = [-10000,-9998,...,-2,1,3,...,9997,9999]) -> [1,4,9,...,100_000_000]', () => {
        const nums: number[] = Array(10000).fill(0)
            .map((_, i) => (i % 2 != 0) ? -(i + 1) : i + 1);
        nums.sort((a, b) => a - b);
        const expected = Array<number>(10000).fill(0).map((_, i) => (i + 1) * (i + 1));

        expect(sortedSquares(nums)).to.eql(expected);
    })

    it('(nums = [-10000,-9998,...,-2,1,3,...,9997,9999]*100) -> [1,4,9,...,100_000_000]*100', () => {
        const repeat = 10;

        let nums: number[] = Array(10000).fill(0)
            .map((_, i) => (i % 2 != 0) ? -(i + 1) : i + 1);
        nums.sort((a, b) => a - b);
        nums = nums.flatMap(n => Array(repeat).fill(n));

        const expected =
            Array(10000).fill(0)
                .map((_, i) => (i + 1) * (i + 1))
                .flatMap((n) => Array(repeat).fill(n))

        expect(sortedSquares(nums)).to.eql(expected);
    })
})

File: src/main/scala/challenge/c2020/c2020_12/c2020_12_15.scala
package challenge.c2020.c2020_12

import org.scalatest.matchers.should.Matchers
import org.scalatest.wordspec.AnyWordSpec

/**
 * [[https://leetcode.com/explore/challenge/card/december-leetcoding-challenge/571/week-3-december-15th-december-21st/3567/]]
 */
//noinspection DuplicatedCode
class c2020_12_15 extends AnyWordSpec with Matchers {

  /**
   * === Squares of a Sorted Array ===
   *
   * Given an integer array `nums` sorted in '''non-decreasing''' order,
   * return ''an array of '''the squares of each number''' sorted in non-decreasing order''.
   *
   * '''Constraints:'''
   *  - `1 <= nums.length <= 10000`
   *  - `-10000 <= nums[i] <= 10000`
   *  - `nums` is sorted in '''non-decreasing''' order.
   */
  object Solution {
    // time O(n); memory O(n)
    def sortedSquares(nums: Array[Int]): Array[Int] = {
      if (nums.length == 1 || nums(0) > 0) nums.map(n => n * n)
      else {
        val result = Array.ofDim[Int](nums.length)

        @scala.annotation.tailrec
        def loop(l: Int, r: Int): Array[Int] = {
          if (l > r) result
          else if (-nums(l) >= nums(r)) {
            result(r - l) = nums(l) * nums(l)
            loop(l + 1, r)
          } else {
            result(r - l) = nums(r) * nums(r)
            loop(l, r - 1)
          }
        }

        loop(0, nums.length - 1)
      }
    }
  }

  // time O(n log n); memory O(n)
  object SolutionScalaBuiltinsBruteForce {
    def sortedSquares(nums: Array[Int]): Array[Int] =
      nums.map(n => n * n).sorted
  }

  import Solution.sortedSquares

  "Example 1: (nums = [-4,-1,0,3,10]) -> [0,1,9,16,100]" in {
    sortedSquares(Array(-4, -1, 0, 3, 10)) shouldBe Array(0, 1, 9, 16, 100)
    //Explanation: After squaring, the array becomes [16,1,0,9,100].
    //After sorting, it becomes [0,1,9,16,100].
  }
  "Example 2: (nums = [-7,-3,2,3,11]) -> [4,9,9,49,121]" in {
    sortedSquares(Array(-7, -3, 2, 3, 11)) shouldBe Array(4, 9, 9, 49, 121)
  }

  "(nums = [-10000,-9998,...,-2,1,3,...,9997,9999]) -> [1,4,9,...,100_000_000]" in {
    val nums = (1 to 10000).map(n => if (n % 2 == 0) -n else n).toArray.sorted
    val expected = (1 to 10000).map(n => n * n).toArray

    sortedSquares(nums) shouldBe expected
  }

  "(nums = [-10000,-9998,...,-2,1,3,...,9997,9999]*100) -> [1,4,9,...,100_000_000]*100" in {
    val repeat = 100

    val nums =
      (1 to 10000)
        .map(n => if (n % 2 == 0) -n else n)
        .flatMap(Array.fill(repeat)(_))
        .toArray
        .sorted
    val expected =
      (1 to 10000)
        .map(n => n * n)
        .flatMap(Array.fill(repeat)(_))
        .toArray

    sortedSquares(nums) shouldBe expected
  }

}
